drivers/can/can_numaker.c

```diff
@@ -0,0 +1,316 @@
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc = 0;
+	uint32_t clksrc_rate_idx;
+	uint32_t clksrc_rate;
+	uint32_t clkdiv_divider;
+
+	/* Initialize clock source rate table
+	 * NOTE: Fix the table for new SOC support
+	 */
+	uint32_t clksrc_rate_tab[] = {
+		__HXT,      /* HXT */
+		0,          /* PLLFOUT/2 */
+		0,          /* HCLK */
+		__HIRC      /* HIRC */
+	};
+	clksrc_rate_tab[1] = CLK_GetPLLClockFreq() / 2;
+	clksrc_rate_tab[2] = CLK_GetHCLKFreq();
+
+	/* SOC in support list
+	 * NOTE: * NOTE: Fix the conditional for new SOC support
+	 */
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_M46X)) {
+		LOG_ERR("Not support SOC: " CONFIG_SOC_SERIES);
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Module clock source rate */
+	clksrc_rate_idx = CLK_GetModuleClockSource(config->clk_modidx);
+	if (clksrc_rate_idx >= ARRAY_SIZE(clksrc_rate_tab)) {
+		LOG_ERR("Invalid clock source rate table or index");
+		rc = -EIO;
+		goto cleanup;
+	}
+	clksrc_rate = clksrc_rate_tab[clksrc_rate_idx];
+
+	/* Module clock divider */
+	clkdiv_divider = CLK_GetModuleClockDivider(config->clk_modidx) + 1;
+
+	LOG_DBG("Clock source/divider: %d/%d", clksrc_rate, clkdiv_divider);
+	*rate = clksrc_rate / clkdiv_divider;
+
+cleanup:
+
+	return rc;
+}
+
+static int can_numaker_init(const struct device *dev)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	struct numaker_scc_subsys scc_subsys;
+	int rc = 0;
+	uint32_t rate;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id        = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx   = config->clk_modidx;
+	scc_subsys.pcc.clk_src      = config->clk_src;
+	scc_subsys.pcc.clk_div      = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	rc = clock_control_on(config->clk_dev, (clock_control_subsys_t) &scc_subsys);
+	if (rc < 0) {
+		goto cleanup;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	rc = clock_control_configure(config->clk_dev, (clock_control_subsys_t)&scc_subsys, NULL);
+	if (rc < 0) {
+		goto cleanup;
+	}
+
+	/* Configure pinmux (NuMaker's SYS MFP) */
+	rc = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (rc < 0) {
+		goto cleanup;
+	}
+
+	/* Reset CAN to default state, same as BSP's SYS_ResetModule(id_rst) */
+	reset_line_toggle_dt(&config->reset);
+
+	config->irq_config_func(dev);
+
+	rc = can_mcan_configure_mram(dev, config->mrba, config->mram);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = can_mcan_init(dev);
+	if (rc < 0) {
+		LOG_ERR("Failed to initialize mcan: %d", rc);
+		goto cleanup;
+	}
+
+	rc = can_numaker_get_core_clock(dev, &rate);
+	if (rc < 0) {
+		goto cleanup;
+	}
+	LOG_INF("CAN core clock: %d", rate);
```

Comment:
Please avoid INF logging in driver init functions. Also, why is this needed? Looks like left-in debugging.

Reply:
Yes, left-in debugging is the better choice.